mempool/src/error/ban_score.rs

```diff
@@ -0,0 +1,300 @@
+            TxIndexError::InvariantBrokenAlreadyUnspent => 0,
+            TxIndexError::InvariantErrorTxNumWrongInBlock(_, _) => 0,
+            TxIndexError::OutputAlreadyPresentInInputsCache => 0,
+            TxIndexError::PreviouslyCachedInputNotFound(_) => 0,
+            TxIndexError::MissingOutputOrSpentOutputErasedOnConnect => 0,
+            TxIndexError::MissingOutputOrSpentOutputErasedOnDisconnect => 0,
+
+            // Invalid transactions
+            TxIndexError::InvalidOutputCount => 100,
+            TxIndexError::SerializationInvariantError(_) => 100,
+            TxIndexError::OutputIndexOutOfRange {
+                tx_id: _,
+                source_output_index: _,
+            } => 100,
+
+            // Double spend may happen if peers are out of sync.
+            // TODO: Punish double spends with many confirmations.
+            TxIndexError::DoubleSpendAttempt(_) => 0,
+            TxIndexError::MissingOutputOrSpent => 0,
+        }
+    }
+}
+
+impl MempoolBanScore for utxo::Error {
+    fn mempool_ban_score(&self) -> u32 {
+        match self {
+            // These errors may be caused by out of sync nodes.
+            // TODO: Make dependent on double spending depth.
+            utxo::Error::UtxoAlreadySpent(_) => 0,
+            utxo::Error::NoUtxoFound => 0,
+
+            // Transaction invalid
+            utxo::Error::InvalidBlockRewardOutputType(_) => 100,
+
+            // Internal errors
+            utxo::Error::OverwritingUtxo => 0,
+            utxo::Error::FreshUtxoAlreadyExists => 0,
+            utxo::Error::NoBlockchainHeightFound => 0,
+            utxo::Error::MissingBlockRewardUndo(_) => 0,
+            utxo::Error::ViewRead => 0,
+            utxo::Error::StorageWrite => 0,
+        }
+    }
+}
+
+impl MempoolBanScore for pos_accounting::Error {
+    fn mempool_ban_score(&self) -> u32 {
+        use pos_accounting::Error as E;
+        match self {
+            // These may be caused by an out of sync peer
+            // TODO(PR): Can we ban the peer in these cases anyway? Some will be only triggered
+            // after an epoch has elapsed which should be enough time for peer to be up-to-date
+            E::InvariantErrorPoolBalanceAlreadyExists => 0,
+            E::InvariantErrorPoolDataAlreadyExists => 0,
+            E::InvariantErrorDelegationCreationFailedIdAlreadyExists => 0,
+            E::AttemptedDecommissionNonexistingPoolBalance => 0,
+            E::AttemptedDecommissionNonexistingPoolData => 0,
+            E::DelegationCreationFailedPoolDoesNotExist => 0,
+            E::DelegateToNonexistingId => 0,
+            E::DelegateToNonexistingPool => 0,
+            E::InvariantErrorPoolCreationReversalFailedBalanceNotFound => 0,
+            E::InvariantErrorPoolCreationReversalFailedDataNotFound => 0,
+            E::InvariantErrorPoolCreationReversalFailedAmountChanged => 0,
+            E::InvariantErrorDelegationShareNotFound => 0,
```

Comment:
What do you think about banning a peer when these occur @azarovh? (See the comment at the top)

Reply:
Two things:

1. Peers that are out of sync are not supposed to relay transactions
2. There's a general problem that has to be handled for PoS, which is that stakes on different branches can appear as double-spent. That will come later.

I'm not sure about these cases specifically and can't do a deep-check it's 3:30 AM here. But I can say that there's some tolerance in some cases based on how bitcoin does this because we cannot guarantee that all peers are at the same tip all the time; there may be slight discrepancy due to network latency or temporary forks.

Reply:
Will leave this as a todo for now and do proper research on how to handle cases where peers may be out of sync. Will be included in a future PR.

Reply:
No problem. Please open an issue referencing, just in case.